datasetinsights/io/gcs.py

```diff
@@ -1,103 +1,227 @@
+import base64
 import logging
 import os
+from os import makedirs
+from os.path import basename, isdir
 from pathlib import Path
 
 from google.cloud.storage import Client
 
+from datasetinsights.io.download import validate_checksum
+from datasetinsights.io.exceptions import (
+    ChecksumError,
+    DownloadError,
+    UploadError,
+)
+
 logger = logging.getLogger(__name__)
 
 
 class GCSClient:
+    """ This class is used to download data from GCS location
+        and perform function such as downloading the dataset and checksum
+        validation.
+    """
+
     def __init__(self, **kwargs):
         """ Initialize a client to google cloud storage (GCS).
         """
         self.client = Client(**kwargs)
 
-    def download(self, bucket_name, object_key, localfile):
-        """ Download a single object from GCS
-        """
-        bucket = self.client.get_bucket(bucket_name)
-        blob = bucket.blob(object_key)
+    def download(self, *, url=None, local_path=None, bucket=None, key=None):
+        """ This method is used to download the dataset from GCS.
 
-        blob.download_to_filename(localfile)
+        Args:
+            url (str): This is the downloader-uri that indicates where
+                              the dataset should be downloaded from.
 
-    def upload(self, localfile, bucket_name, object_key):
-        """ Upload a single object to GCS
-        """
-        bucket = self.client.get_bucket(bucket_name)
-        blob = bucket.blob(object_key)
+            local_path (str): This is the path to the directory where the
+                          download will store the dataset.
 
-        blob.upload_from_filename(localfile)
+            bucket (str): gcs bucket name
+            key (str): object key path
 
+            Examples:
+                >>> url = "gs://bucket/folder or gs://bucket/folder/data.zip"
+                >>> local_path = "/tmp/folder"
+                >>> bucket ="bucket"
+                >>> key ="folder/data.zip" or "folder"
 
-def gcs_bucket_and_path(url):
-    """Split an GCS-prefixed URL into bucket and path."""
-    gcs_prefix = "gs://"
-    if not url.startswith(gcs_prefix):
-        raise ValueError(
-            f"Specified destination prefix: {url} does not start "
-            f"with {gcs_prefix}"
-        )
-    url = url[len(gcs_prefix) :]
-    idx = url.index("/")
-    bucket = url[:idx]
-    path = url[(idx + 1) :]
+        """
+        if not (bucket and key) and url:
+            bucket, key = self._parse(url)
 
-    return bucket, path
+        bucket_obj = self.client.get_bucket(bucket)
+        if self._is_file(bucket_obj, key):
+            self._download_file(bucket_obj, key, local_path)
+        else:
+            self._download_folder(bucket_obj, key, local_path)
 
+    def _download_folder(self, bucket, key, local_path):
+        """ download all files from directory
+        """
+        blobs = bucket.list_blobs(prefix=key)
```

Comment:
Will `bucket.list_blobs` recursively go through the GCS storage hierarchy? Will `_downliad_folder` download all files recursively?

Reply:
There is no folder hierarchy on GCS, the are only keys file files. I think the prefix catches all that come after that prefix which is similar to a folder structure.